src/common/src/util/value_encoding/mod.rs

```diff
@@ -15,94 +15,204 @@
 use crate::types::{
-    deserialize_datum_from, deserialize_datum_not_null_from, serialize_datum_into,
-    serialize_datum_not_null_into, DataType, Datum, Decimal, ScalarImpl,
+    DataType, Datum, Decimal, IntervalUnit, NaiveDateTimeWrapper, NaiveDateWrapper,
+    NaiveTimeWrapper, OrderedF32, OrderedF64, ScalarImpl, ScalarRefImpl,
 };
 
 /// Serialize datum into cell bytes (Not order guarantee, used in value encoding).
 pub fn serialize_cell(cell: &Datum) -> Result<Vec<u8>> {
-    let mut serializer = value_encoding::Serializer::new(vec![]);
-    if let Some(ScalarImpl::Decimal(decimal)) = cell {
-        return serialize_decimal(decimal);
+    let mut buf: Vec<u8> = vec![];
+    if let Some(datum) = cell {
+        buf.put_u8(1);
+        serialize_datum(datum.as_scalar_ref_impl(), &mut buf)
+    } else {
+        buf.put_u8(0);
     }
-    serialize_datum_into(cell, serializer.memcom_ser())?;
-    Ok(serializer.into_inner())
+    Ok(buf)
 }
 
 /// Serialize datum cannot be null into cell bytes.
 pub fn serialize_cell_not_null(cell: &Datum) -> Result<Vec<u8>> {
-    let mut serializer = value_encoding::Serializer::new(vec![]);
-    if let Some(ScalarImpl::Decimal(decimal)) = cell {
-        return serialize_decimal(decimal);
-    }
-    serialize_datum_not_null_into(cell, serializer.memcom_ser())?;
-    Ok(serializer.into_inner())
+    let mut buf: Vec<u8> = vec![];
+    serialize_datum(
+        cell.as_ref()
+            .expect("datum cannot be null")
+            .as_scalar_ref_impl(),
+        &mut buf,
+    );
+    Ok(buf)
 }
 
 /// Deserialize cell bytes into datum (Not order guarantee, used in value decoding).
-pub fn deserialize_cell(
-    deserializer: &mut value_encoding::Deserializer<impl Buf>,
-    ty: &DataType,
-) -> Result<Datum> {
-    match ty {
-        &DataType::Decimal => deserialize_decimal(deserializer),
-        _ => Ok(deserialize_datum_from(ty, deserializer.memcom_de())?),
+pub fn deserialize_cell<B: Buf>(data: &mut B, ty: &DataType) -> Result<Datum> {
+    let null_tag = data.get_u8();
+    match null_tag {
+        0 => {
+            return Ok(None);
+        }
+        1 => {}
+        _ => {
+            return Err(RwError::from(ErrorCode::InternalError(format!(
+                "Invalid null tag: {}",
+                null_tag
+            ))));
+        }
     }
+    Ok(deserialize_datum(ty, data)?)
 }
 
 /// Deserialize cell bytes which cannot be null into datum.
-pub fn deserialize_cell_not_null(
-    deserializer: &mut value_encoding::Deserializer<impl Buf>,
-    ty: DataType,
-) -> Result<Datum> {
-    match ty {
-        DataType::Decimal => deserialize_decimal(deserializer),
-        _ => Ok(deserialize_datum_not_null_from(
-            ty,
-            deserializer.memcom_de(),
-        )?),
+pub fn deserialize_cell_not_null<B: Buf>(data: &mut B, ty: &DataType) -> Result<Datum> {
+    Ok(deserialize_datum(ty, data)?)
+}
+
+fn serialize_datum(value: ScalarRefImpl, buf: &mut Vec<u8>) {
```

Comment:
`serialize_datum` might be easily confused with memcomprable's serialize_datum. Better to change a name here.

Reply:
it's ok. or maybe always use value_encoding::serialize_datum